src/redis_cache.h

```diff
@@ -0,0 +1,130 @@
+//    notice, this list of conditions and the following disclaimer in the
+//    documentation and/or other materials provided with the distribution.
+//  * Neither the name of NVIDIA CORPORATION nor the names of its
+//    contributors may be used to endorse or promote products derived
+//    from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS ``AS IS'' AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+// OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+#pragma once
+
+#include <list>
+#include <string>
+#include <iostream>
+#include <string_view>
+#include <unordered_map>
+
+#include "triton/common/model_config.h"
+#include "triton/common/logging.h"
+#include "infer_stats.h"
+#include "response_cache.h"
+#include <sw/redis++/redis++.h>
+
+
+template <typename Tkv>
+struct HashEntry {
+  std::unordered_map<Tkv, Tkv> fields;
+};
+
+template <typename Tkv>
+struct RedisCacheEntry {
+  std::string key;
+  int num_entries = 1;
+  std::vector<HashEntry<Tkv>> outputs;
+};
+
+
+
+namespace triton { namespace core {
+
+
+class RedisResponseCache : public RequestResponseCache {
+ public:
+  ~RedisResponseCache();
+
+  // Create the request/response cache object
+  static Status Create(std::string address, std::string username, std::string password, std::unique_ptr<RequestResponseCache>* cache);
+
+  Status Lookup(
+      InferenceResponse* const response, InferenceRequest* const request);
+
+  Status Insert(
+      const InferenceResponse& response, InferenceRequest* const request);
+
+  bool Exists(const uint64_t key);
+
+  // dummy method to statisfy compiler
+  // there is bound to be a better way to do this.
+  Status Evict() {
+    Status status = Status(Status::Code::SUCCESS);
+    return status;
+  }
+
+  // Returns number of items in cache
+  size_t NumEntries()
+  {
+    // TODO consider how this will work with dual key design
+    // will currently return 2x number of keys by design
+    return (size_t)_client->dbsize();
+  }
+  // Returns number of items evicted in cache lifespan
+  size_t NumEvictions()
+  {
+    // TODO get from Redis instead
+    return RequestResponseCache::NumEvictions();
+  }
+
+  // TODO Are these needed for Redis? It's easy to get the values
+  // Returns total number of bytes allocated for cache
```

Comment:
Triton exposes various Cache metrics described here: https://github.com/triton-inference-server/server/blob/main/docs/user_guide/metrics.md#response-cache-metrics

It would be nice to satisfy as many of these existing metrics as possible for the Redis version. 

Also if I recall correctly, the dashboards provided by Redis for observability only apply to the Enterprise version. These metrics may be especially valuable to Triton users making use of the OSS Redis version who don't have access to these dashboards.

Reply:
the "INFO" command exposes all of this information and making ``get`` methods is easy enough. OSS or enterprise, doesn't matter. it's the same API.

heres a sample of some of the fields (for OSS or enterprise)

```
redis_version:4.0.9
redis_git_sha1:00000000
redis_git_dirty:0
redis_build_id:9435c3c2879311f3
redis_mode:standalone
os:Linux 4.15.0-192-generic x86_64
arch_bits:64
multiplexing_api:epoll
atomicvar_api:atomic-builtin
gcc_version:7.4.0
process_id:1649
run_id:0aa36241acbb2bc1cff081b05a757dbe604f31a3
tcp_port:6379
uptime_in_seconds:1906726
uptime_in_days:22
hz:10
lru_clock:4675455
executable:/usr/bin/redis-server
config_file:/etc/redis/redis.conf

# Clients
connected_clients:3
client_longest_output_list:0
client_biggest_input_buf:0
blocked_clients:0

# Memory
used_memory:903264
used_memory_human:882.09K
used_memory_rss:4018176
used_memory_rss_human:3.83M
used_memory_peak:939064
used_memory_peak_human:917.05K
used_memory_peak_perc:96.19%
used_memory_overhead:866946
used_memory_startup:782504
used_memory_dataset:36318
used_memory_dataset_perc:30.07%
total_system_memory:33579556864
total_system_memory_human:31.27G
used_memory_lua:37888
used_memory_lua_human:37.00K
maxmemory:0
maxmemory_human:0B
maxmemory_policy:noeviction
mem_fragmentation_ratio:4.45
mem_allocator:jemalloc-3.6.0
active_defrag_running:0
lazyfree_pending_objects:0

# Persistence
loading:0
rdb_changes_since_last_save:0
rdb_bgsave_in_progress:0
rdb_last_save_time:1665619127
rdb_last_bgsave_status:ok
rdb_last_bgsave_time_sec:0
rdb_current_bgsave_time_sec:-1
rdb_last_cow_size:315392
aof_enabled:0
aof_rewrite_in_progress:0
aof_rewrite_scheduled:0
aof_last_rewrite_time_sec:-1
aof_current_rewrite_time_sec:-1
aof_last_bgrewrite_status:ok
aof_last_write_status:ok
aof_last_cow_size:0

# Stats
total_connections_received:200
total_commands_processed:1704
instantaneous_ops_per_sec:0
total_net_input_bytes:176779
total_net_output_bytes:555763
instantaneous_input_kbps:0.00
instantaneous_output_kbps:0.00
rejected_connections:0
sync_full:0
sync_partial_ok:0
sync_partial_err:0
expired_keys:0
expired_stale_perc:0.00
expired_time_cap_reached_count:0
evicted_keys:0
keyspace_hits:139
keyspace_misses:44
pubsub_channels:0
pubsub_patterns:0
latest_fork_usec:320
migrate_cached_sockets:0
slave_expires_tracked_keys:0
active_defrag_hits:0
active_defrag_misses:0
active_defrag_key_hits:0
active_defrag_key_misses:0

# Replication
role:master
connected_slaves:0
master_replid:71ce9bae2f8ef2da1bb814d06365c73bf10ac5af
master_replid2:0000000000000000000000000000000000000000
master_repl_offset:0
second_repl_offset:-1
repl_backlog_active:0
repl_backlog_size:1048576
repl_backlog_first_byte_offset:0
repl_backlog_histlen:0

# CPU
used_cpu_sys:2081.27
used_cpu_user:749.44
used_cpu_sys_children:0.06
used_cpu_user_children:0.07

# Cluster
cluster_enabled:0

# Keyspace
db0:keys=21,expires=0,avg_ttl=0
```